### PR TITLE
Add Presentation

### PR DESCRIPTION
**Pull request by Joseph-Edwards**
This PR includes the `Presentation` class in a state ready for V3. In particular:
- [X] use of `no_checks` functions
- [X] TODOs and FIXMEs reviewed
- [x] doc
- [x] Python bindings
- [x] `iwyu`

### CODE REVIEW

**Comment by Joseph-Edwards** [comment]
This is a draft at the moment; I have a few quick questions for @james-d-mitchell before I'd be happy to merge.

**Comment by Joseph-Edwards** [comment]
@james-d-mitchell the changes we made to get rid of `enable_if_t` seemingly didn't work when running `test_all`. I've changed it to the `auto foo() -> std::enable_if_t<...>` thing at that's working now.

**Comment by Joseph-Edwards** [comment]
It's worth noting that this PR doesn't particularly address `InversePresentations`, and they haven't been added to the bindings either. Should this be a part of this PR? Or a separate one?

**Comment by Joseph-Edwards** [comment]
Making this a draft whilst I add tests for `InversePresentation`

**Comment by Joseph-Edwards** [comment]
I have no idea why the Azure tests are failing. The failing test is `[029]` in `test-cong.hpp`, and this test seems to have been failing on the CI for a while. I can't reproduce this error, and I'm not really sure how to resolve it.

File: include/libsemigroups/presentation.tpp [hunk 1]
```diff
@@ -185,6 +185,33 @@ namespace libsemigroups {
...
+      typename Presentation<Word>::letter_type x) {
+    if (in_alphabet(x)) {
+      remove_generator_no_checks(x);
+    } else {
+      LIBSEMIGROUPS_EXCEPTION("the argument {} does not belong to the alphabet "
+                              "{}, expected an existing letter",
+                              detail::to_printable(x),
+                              detail::to_printable(alphabet()));
+    }
+  }
```
Comment (Joseph-Edwards) [inline include/libsemigroups/presentation.tpp:213]: What should the checks for this function be? The current option is the bare minimum; check `x` is in the alphabet. However, if `x` is duplicated in the alphabet, then strange things can happen. I wasn't sure where to place this on the "let the user do what they want" to "don't let the user break things" scale.

File: include/libsemigroups/presentation.hpp [hunk 2]
```diff
@@ -27,10 +27,12 @@
 
 #include <algorithm>         // for reverse, sort
 #include <cstring>           // for size_t, strlen
+#include <ctype.h>           // for isprint
 #include <initializer_list>  // for initializer_list
 #include <iterator>          // for distance
 #include <limits>            // for numeric_limits
 #include <map>               // for map
+#include <math.h>            // for pow
```
Comment (james-d-mitchell) [inline include/libsemigroups/presentation.hpp:35]:
```suggestion
#include <cmath>            // for pow
```

File: include/libsemigroups/presentation.tpp (hunk 1 above)
Comment (james-d-mitchell) [reply include/libsemigroups/presentation.tpp]: Is this still an issue?

Comment (Joseph-Edwards) [reply include/libsemigroups/presentation.tpp]: @james-d-mitchell yes, still not sure what the best course of action is. I think the most sensible option is to check that the alphabet is valid; the user should only be able to significantly mess up the internals of a presentation through a `_no_checks` function. What do you think?

File: include/libsemigroups/presentation.tpp [hunk 3]
```diff
@@ -918,23 +972,166 @@ namespace libsemigroups {
...
+    }
+    out += ">";
+    return out;
+  }
+
+  template <typename Word>
+  std::string to_human_readable_repr(InversePresentation<Word> const& p) {
+    size_t      alphabet_size = p.alphabet().size();
+    size_t      n_rules       = p.rules.size() / 2;
+    std::string out           = "<";
```
Comment (james-d-mitchell) [inline include/libsemigroups/presentation.tpp:1098]:
```suggestion
    std::string out           = "<inverse ";
```

File: include/libsemigroups/presentation.tpp [hunk 4]
```diff
@@ -918,23 +972,166 @@ namespace libsemigroups {
...
+    if (p.contains_empty_word()) {
+      out += "Monoid ";
+    } else {
+      out += "Semigroup ";
+    }
+    out += "Presentation with inverses on ";
```
Comment (james-d-mitchell) [inline include/libsemigroups/presentation.tpp:1104]:
```suggestion
    out += "presentation with ";
```

File: include/libsemigroups/presentation.tpp [hunk 5]
```diff
@@ -918,23 +972,166 @@ namespace libsemigroups {
...
+    if (alphabet_size == 1) {
```
Comment (james-d-mitchell) [inline include/libsemigroups/presentation.tpp:1105]:
```suggestion
    fmt::format("<inverse {} presentation with {} letter{}, {} rule{}, and length {}>", 
    (p.contains_empty_word() ? "monoid" : "semigroup"),
    alphabet_size, 
    (alphabet_size > 1 ? "s" : ""),
    n_rules,
    (n_rules > 1 ? "s" : ""),
    presentation::length(p));
```

File: include/libsemigroups/presentation.tpp [hunk 6]
```diff
@@ -918,23 +972,166 @@ namespace libsemigroups {
...
+    return *this;
+  }
+
+  template <typename Word>
+  typename InversePresentation<Word>::letter_type
+  InversePresentation<Word>::inverse(letter_type x) const {
+    if (_inverses.empty()) {
+      LIBSEMIGROUPS_EXCEPTION("no inverses have been defined")
+    }
+    return _inverses[this->index(x)];
```
Comment (james-d-mitchell) [inline include/libsemigroups/presentation.tpp:1135]:
```suggestion
    return _inverses[Presentation<Word>::index(x)];
```

File: include/libsemigroups/presentation.tpp [hunk 7]
```diff
@@ -1067,53 +1067,23 @@ namespace libsemigroups {
...
-    out += "with ";
-    if (n_rules == 1) {
-      out += "1 rule";
-    } else {
-      out += std::to_string(n_rules) + " rules";
-    }
-    out += ">";
+    std::string out
+        = to_human_readable_repr(static_cast<Presentation<Word> const&>(p));
+    out.insert(1, "inverse ");
```
Comment (Joseph-Edwards) [inline include/libsemigroups/presentation.tpp:1086]: @james-d-mitchell is this the best way to call a parent class version of a free function? It feels like there should be a nicer, more inheritance-like way of doing this.

Comment (james-d-mitchell) [reply include/libsemigroups/presentation.tpp:1086]: Looks good to me

**Comment by Joseph-Edwards** [comment]
@james-d-mitchell I think this is ready to go now if you're happy with it.

**Pull request merged**